Add blog for presto-hudi deep dive (#200)
prestodb/prestodb.github.io: website/blog/2023-04-04-presto-hudi-deep-dive.md

---
author: Pratyaksh Sharma
authorURL: https://www.linkedin.com/in/pratyaksh-sharma-60593769/
title: Presto Hudi Deep Dive
---

> With the growing popularity of the lakehouse approach, it has become increasingly important for query engines to support these new formats such as Hudi. A [previous blog](https://prestodb.io/blog/2020/08/04/prestodb-and-hudi) discusses the evolution of presto-hudi integration via hive connector at a high level. With the latest community developments, a [separate presto-hudi connector](https://github.com/prestodb/presto/issues/17006) has come up but it is not at par with the hive connector in terms of security features, caching and cost-based optimization. In this blog, we dive deeper into presto-hudi integration supported via hive connector from a developer's perspective and list down the various checkpoints where the query execution moves from presto to hudi library.

Reviewer: Can you elaborate more on "caching and cost-based optimization" of Hive connector? Maybe, ti's something we can add in Hudi connector too. We already leverage metastore caching.

Author: Sure, will add single liners.

<!--truncate-->

## Input formats and Record readers
Before actually diving into the presto-hudi integration, let us first review a few related concepts. In the Hadoop world, data is stored in a logical file. This file can have various physical layouts (format) on remote storage. A "Path" identifies a file. Further, an "input format" abstraction allows reading portions (splits) of a file. To keep it short and sweet, input formats help generate splits from files which are logical representations of the data. Input formats provide record readers which do the actual reading from these splits.

InputFormat interface exposes [2 methods](https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/mapred/InputFormat.html) - `getSplits()` and `getRecordReader()`.

### Creation of splits

> *`getSplits()`* method facilitates the creation of splits from actual data files. On a high level, this involves reading the input paths from the configuration object and scanning those paths to get the files present. Files are logically represented as FileStatus. These file paths are then passed to the `makeSplit()` method which creates the splits for further processing. Below are the snippets of the actual code where these calls are being made.

Reviewer: no need to italicize

![](/img/blog/2023-04-04-presto-hudi-deep-dive/get-splits-call.png)
*Figure 1: listStatus() gets called internally from getSplits().*

![](/img/blog/2023-04-04-presto-hudi-deep-dive/make-split-call.png)
*Figure 2: makeSplit() call is highlighted as part of getSplits() implementation.*

> The above snippets are taken from FileInputFormat class. Please note that the lines are highlighted where these calls are being made. 

Reviewer: Can also link to the `FileInoutFormat` code.

> The above snippets are taken from FileInputFormat class. Please note that the lines are highlighted where these calls are being made. 

Reviewer: ```suggestion
The above snippets are taken from `FileInputFormat` class. Please note that the lines are highlighted where these calls are being made. 
```

> ![](/img/blog/2023-04-04-presto-hudi-deep-dive/get-splits-call.png)

Reviewer: Would it be better to share a permalink to the code instead of screenshots?

Author: I would prefer adding permalink along with having screenshot here. The reason is with code changes, permalink can start pointing to some other piece of code. Will do the needful.

## Hudi table types

> Hudi supports 2 types of tables out of the box, namely, Copy on Write (CoW) and Merge on Read (MoR). CoW is the simpler of the two in terms of the reading and writing complexities involved. Upserts made to this table type result in newer versions of columnar parquet files which can be read using the ParquetPageSource used internally by Presto.

Reviewer: ```suggestion
Hudi supports two types of tables, namely, Copy on Write (CoW) and Merge on Read (MoR). CoW is the simpler of the two in terms of the reading complexities involved. Upserts made to this table type result in newer versions of columnar parquet files which can be read efficiently using the native `ParquetPageSource` used internally by Presto.
```

> Hudi supports 2 types of tables out of the box, namely, Copy on Write (CoW) and Merge on Read (MoR). CoW is the simpler of the two in terms of the reading and writing complexities involved. Upserts made to this table type result in newer versions of columnar parquet files which can be read using the ParquetPageSource used internally by Presto.

Reviewer: actually, writing CoW is more complex than it seems to be. But, while reading it's much easier as there is no merge involved.

Author: Yeah right. I was actually referring to end user perspective. For end users, they only see a newer version of parquet file written with CoW. With MoR, they see multiple files written with delta commits. 
Anyways I will change it as per your suggestion.

> Hudi supports 2 types of tables out of the box, namely, Copy on Write (CoW) and Merge on Read (MoR). CoW is the simpler of the two in terms of the reading and writing complexities involved. Upserts made to this table type result in newer versions of columnar parquet files which can be read using the ParquetPageSource used internally by Presto.

Reviewer: can also link to https://hudi.apache.org/docs/table_types

Author: Thank you for the suggestion.

> MoR involves writing incoming upserts to delta log files in avro format by default. These log files are later compacted to create new base parquet files. Since this involves row and columnar format files, this table type supports Read Optimized (RO) and Real Time (RT) queries.

Reviewer: ```suggestion
MoR involves writing incoming updates to delta log files in avro format by default. These log files are later compacted to create new base parquet files. Since this involves row and columnar format files, this table type supports Read Optimized (RO) and Real Time (RT) queries.
```

Author: Thank you for pointing this out.

From Presto’s point of view, CoW table and `_ro` version of a MoR table behave the same way. However `_rt` version of MoR table type requires additional work as described in the next section.

![](/img/blog/2023-04-04-presto-hudi-deep-dive/hudi-mor-ro-1.png)

![](/img/blog/2023-04-04-presto-hudi-deep-dive/hudi-mor-ro-2.png)
*Figure 3: Snapshot showing details about _ro table.*

![](/img/blog/2023-04-04-presto-hudi-deep-dive/hudi-mor-rt-1.png)

![](/img/blog/2023-04-04-presto-hudi-deep-dive/hudi-mor-rt-2.png)
*Figure 4: Snapshot showing details about _rt table.*

The above figures show the difference between the 2 versions of MoR tables in terms of table properties and input formats involved.

As is clearly visible, Hudi syncs the CoW and `_ro` tables with `HoodieParquetInputFormat` while `_rt` tables are registered with `HoodieParquetRealtimeInputFormat`. 

### Merge on Read table type

> Currently presto supports read optimized and real time queries for MoR tables via the hive connector. After the parser, planner and optimizer phases, splits are generated by Presto which are used for creating presto pages to be rendered back to the client. We focus on this split generation phase and the record reader involved thereafter in the next sections. 

Reviewer: real time queries are also supported in Hudi connector - https://github.com/prestodb/presto/blob/0c08b8db0077a56636b9e0ddd6d4036eb3178a8c/presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java#L110

Author: Right, actually the focus is only on hudi support via hive connector.

Reviewer: The title of the blog suggests that it is Presto and Hudi deep dive overall. Maybe we can make it clear in the title, something like "Hudi Tables on Presto-Hive connector: A Deep Dive"?

> Currently presto supports read optimized and real time queries for MoR tables via the hive connector. After the parser, planner and optimizer phases, splits are generated by Presto which are used for creating presto pages to be rendered back to the client. We focus on this split generation phase and the record reader involved thereafter in the next sections. 

Reviewer: ```suggestion
Currently presto supports read optimized and real time queries for MoR tables via the hive as well as the hudi connector. After the parser, planner and optimizer phases, splits are generated by Presto which are used for creating presto pages to be rendered back to the client. We focus on this split generation phase and the record reader involved thereafter in the next sections. 
```

Author: Do you want me to do this change? At other places in this blog, I am only talking about hudi support via hive connector. I feel adding it explicitly for MoR tables might result in misunderstandings that hudi connector is lacking in other features that we are mentioning, for example, CoW reading etc. Happy to hear your thoughts on this @codope

Reviewer: Makes sense. Let's keep it the way you have written but let's also make it clear in the title as suggested above. We can even remove `via the hive connector` part.

Author: sure, that should make it more clear

Now refer to Figure 1, `listStatus()` is called which is overridden in `HoodieCopyOnWriteTableInputFormat.listStatus()`. This method in turn calls `listStatusForIncrementalMode()` and `listStatusForSnapshotMode()`. This will return `RealtimeFileStatus`. `RealtimeFileStatus` overrides `getPath()` method which returns `HoodieRealtimePath`. Please refer to the below figures -  

![](/img/blog/2023-04-04-presto-hudi-deep-dive/list-status-mor.png)
*Figure 7: listStatus() call for MoR table type returns RealtimeFileStatus.*

![](/img/blog/2023-04-04-presto-hudi-deep-dive/make-split-mor.png)

*Figure 8: makeSplit() internally called from getSplits() for MoR table type returns HoodieRealtimeFileSplit.*

![](/img/blog/2023-04-04-presto-hudi-deep-dive/get-path-mor.png)

*Figure 9: getPath() is overridden in RealtimeFileStatus and it returns HoodieRealtimePath.*

> Now that real time splits got generated, we need a way of reading the records from these special splits by merging the base and delta files. `GenericHiveRecordCursorProvider` calls createRecordReader which ultimately calls `inputFormat.getRecordReader()`. This returns `RealtimeCompactedRecordReader` by default. Please refer to the below figure - 

Reviewer: Should we also explain the merging of records, or at the very least, why we're using the record reader as we need  custom reader that understand Hudi MOR format?

Author: Sure.

![](/img/blog/2023-04-04-presto-hudi-deep-dive/real-time-record-reader.png)
*Figure 10: Real time record reader.*

#### Read Optimized Queries

These queries will occur from `<table_name>_ro` table registered in HMS and it has HoodieParquetInputFormat as the input format. These queries will only read the base parquet files, hence the name read optimized. The execution of these queries is exactly similar to that for CoW table type. Please refer to the next section for more details.

### Copy on Write table type

> The layout of this table type only consists of parquet files. This makes use of the HudiDirectoryLister class for listing relevant data files. HudiDirectoryLister is a singleton object and gets created once every query involving hudi CoW table. This class creates the hudi timeline by scanning the .hoodie folder, then creates a file system view on top of this timeline. FileSystemView is a logical representation of the hudi storage layout which binds the timeline and data files together.

Reviewer: ```suggestion
The layout of this table type only consists of parquet files. This makes use of the `HudiDirectoryLister` class for listing relevant data files. `HudiDirectoryLister` is a singleton object and gets created once every query involving hudi CoW table. This class creates the hudi timeline by scanning the `.hoodie` folder, then creates a file system view on top of this timeline. `FileSystemView` is a logical representation of the hudi storage layout which binds the timeline and data files together.
```

> Unlike HadoopDirectoryLister, this lister does not make use of the file listing cache since the hudi table is assumed to be ever evolving and hence the partitions are not sealed. Once the splits are created, native ParquetPageSource is used for rendering the results.

Reviewer: ```suggestion
Unlike `HadoopDirectoryLister`, this lister does not make use of the file listing cache since the hudi table is assumed to be ever-evolving and hence the partitions are not sealed. Once the splits are created, native `ParquetPageSource` is used for rendering the results.
```
